src/Components/WebAssembly/WebAssembly.Authentication/src/Services/AuthorizationMessageHandler.cs

```diff
@@ -0,0 +1,124 @@
+    /// <see cref="ConfigureHandler(IEnumerable{string}, IEnumerable{string}, string)"/>.
+    /// </summary>
+    public class AuthorizationMessageHandler : DelegatingHandler
+    {
+        private readonly IAccessTokenProvider _provider;
+        private readonly NavigationManager _navigation;
+        private AccessToken _lastToken;
+        private AuthenticationHeaderValue _cachedHeader;
+        private Uri[] _allowedUris;
+        private AccessTokenRequestOptions _tokenOptions;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="AuthorizationMessageHandler"/>.
+        /// </summary>
+        /// <param name="provider">The <see cref="IAccessTokenProvider"/> to use for provisioning tokens.</param>
+        /// <param name="navigation">The <see cref="NavigationManager"/> to use for performing redirections.</param>
+        public AuthorizationMessageHandler(
+            IAccessTokenProvider provider,
+            NavigationManager navigation)
+        {
+            _provider = provider;
+            _navigation = navigation;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var now = DateTimeOffset.Now;
+            if (_allowedUris == null)
+            {
+                throw new InvalidOperationException($"The '{nameof(AuthorizationMessageHandler)}' is not configured. " +
+                    $"Call '{nameof(AuthorizationMessageHandler.ConfigureHandler)}' and provide a list of endpoint urls to attach the token to.");
+            }
+
+            if (_allowedUris.Any(uri => uri.IsBaseOf(request.RequestUri)))
+            {
+                if (_lastToken == null || now >= _lastToken.Expires.AddMinutes(-5))
+                {
+                    var tokenResult = _tokenOptions != null ?
+                        await _provider.RequestAccessToken(_tokenOptions) :
+                        await _provider.RequestAccessToken();
+
+                    if (tokenResult.TryGetToken(out var token))
+                    {
+                        _lastToken = token;
+                        _cachedHeader = new AuthenticationHeaderValue("Bearer", _lastToken.Value);
+                    }
+                    else
+                    {
+                        throw new AccessTokenNotAvailableException(tokenResult, _navigation);
+                    }
+                }
+
+                // We don't try to handle 401s and retry the request with a new token automatically since that would mean we need to copy the request
+                // headers and buffer the body and we expect that the user instead handles the 401s. (Also, we can't really handle all 401s as we might
+                // not be able to provision a token without user interaction).
+                request.Headers.Authorization = _cachedHeader;
```

Comment:
I know we're in the realms of pure theory here, but in theory there's a thread safety issue: `_cachedHeader` might still be null after `_lastToken` was assigned on another thread.

Where possible it's good to account for thready safety even on Blazor WebAssembly now as we want to be enabling the threaded version of `dotnet.wasm` for 5.0.

Reply:
Filed a separate issue for it. We don't need it right now and I don't want it to creep into the scope of this PR. https://github.com/dotnet/aspnetcore/issues/20726